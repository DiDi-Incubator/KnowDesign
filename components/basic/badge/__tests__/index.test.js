import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Badge from '../index';
import Tooltip from '../../tooltip';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';

describe('Badge', () => {
  mountTest(Badge);
  rtlTest(Badge);
  rtlTest(() => (
    <Badge count={5} offset={[10, 10]}>
      <a href="#" className="head-example">
        head
      </a>
    </Badge>
  ));

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('badge dot not scaling count > 9', () => {
    const badge = mount(<Badge count={10} dot />);
    expect(badge.find('.ant-card-multiple-words').length).toBe(0);
  });

  it('badge should support float number', () => {
    let wrapper = mount(<Badge count={3.5} />);
    expect(wrapper.find('.ant-badge-multiple-words').first().text()).toEqual('3.5');

    wrapper = mount(<Badge count="3.5" />);
    expect(wrapper.find('.ant-badge-multiple-words').first().text()).toEqual('3.5');
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('badge dot not showing count == 0', () => {
    const badge = mount(<Badge count={0} dot />);
    expect(badge.find('.ant-badge-dot').length).toBe(0);
  });

  it('should have an overriden title attribute', () => {
    const badge = mount(<Badge count={10} title="Custom title" />);
    expect(
      badge.find('.ant-scroll-number').getDOMNode().attributes.getNamedItem('title').value,
    ).toEqual('Custom title');
  });

  // https://github.com/ant-design/ant-design/issues/10626
  it('should be composable with Tooltip', () => {
    const ref = React.createRef();
    const wrapper = mount(
      <Tooltip title="Fix the error" ref={ref}>
        <Badge status="error" />
      </Tooltip>,
    );

    act(() => {
      wrapper.find('Badge').simulate('mouseenter');
      jest.runAllTimers();
    });
    expect(ref.current.props.visible).toBeTruthy();
  });

  it('should render when count is changed', () => {
    const wrapper = mount(<Badge count={9} />);

    function updateMatch(count) {
      wrapper.setProps({ count });

      act(() => {
        jest.runAllTimers();
        wrapper.update();
        expect(wrapper.render()).toMatchSnapshot();
      });
    }

    updateMatch(10);
    updateMatch(11);
    updateMatch(11);
    updateMatch(111);
    updateMatch(10);
    updateMatch(9);
  });

  it('should be compatible with borderColor style', () => {
    const wrapper = mount(
      <Badge
        count={4}
        style={{ backgroundColor: '#fff', color: '#999', borderColor: '#d9d9d9' }}
      />,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/13694
  it('should support offset when count is a ReactNode', () => {
    const wrapper = mount(
      <Badge count={<span className="custom" style={{ color: '#f5222d' }} />} offset={[10, 20]}>
        <a href="#" className="head-example">
          head
        </a>
      </Badge>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/15349
  it('should color style  works on Badge', () => {
    const wrapper = mount(<Badge style={{ color: 'red' }} status="success" text="Success" />);
    expect(wrapper.find('.ant-badge-status-text').props().style.color).toBe('red');
  });

  // https://github.com/ant-design/ant-design/issues/15799
  it('render correct with negative number', () => {
    const wrapper = mount(
      <div>
        <Badge count="-10" />
        <Badge count={-10} />
      </div>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/21331
  // https://github.com/ant-design/ant-design/issues/31590
  it('render Badge status/color when contains children', () => {
    const wrapper = mount(
      <div>
        <Badge count={5} status="success">
          <a />
        </Badge>
        <Badge count={5} color="blue">
          <a />
        </Badge>
        <Badge count={5} color="#08c">
          <a />
        </Badge>
      </div>,
    );
    expect(wrapper.render()).toMatchSnapshot();
    expect(wrapper.find(Badge).at(0).find('.ant-scroll-number-only-unit').text()).toBe('5');
    expect(wrapper.find(Badge).at(1).find('.ant-scroll-number-only-unit').text()).toBe('5');
    expect(wrapper.find(Badge).at(2).find('.ant-scroll-number-only-unit').text()).toBe('5');
  });

  it('Badge should work when status/color is empty string', () => {
    const wrapper = mount(
      <>
        <Badge color="" text="text" />
        <Badge status="" text="text" />
      </>,
    );

    expect(wrapper.find('.ant-badge')).toHaveLength(2);
  });
});
