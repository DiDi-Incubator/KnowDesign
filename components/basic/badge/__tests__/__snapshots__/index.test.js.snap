// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge render Badge status/color when contains children 1`] = `
LoadedCheerio {
  "0": <div>
    <span
      class="dcloud-badge dcloud-badge-status"
    >
      <a />
      <sup
        class="dcloud-scroll-number dcloud-badge-count dcloud-badge-status-success"
        data-show="true"
        title="5"
      >
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            5
          </span>
        </span>
      </sup>
    </span>
    <span
      class="dcloud-badge dcloud-badge-status"
    >
      <a />
      <sup
        class="dcloud-scroll-number dcloud-badge-count dcloud-badge-status-blue"
        data-show="true"
        title="5"
      >
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            5
          </span>
        </span>
      </sup>
    </span>
    <span
      class="dcloud-badge dcloud-badge-status"
    >
      <a />
      <sup
        class="dcloud-scroll-number dcloud-badge-count"
        data-show="true"
        style="background: rgb(0, 136, 204);"
        title="5"
      >
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            5
          </span>
        </span>
      </sup>
    </span>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge render correct with negative number 1`] = `
LoadedCheerio {
  "0": <div>
    <span
      class="dcloud-badge dcloud-badge-not-a-wrapper"
    >
      <sup
        class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
        data-show="true"
        title="-10"
      >
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            -
          </span>
        </span>
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            1
          </span>
        </span>
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            0
          </span>
        </span>
      </sup>
    </span>
    <span
      class="dcloud-badge dcloud-badge-not-a-wrapper"
    >
      <sup
        class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
        data-show="true"
        title="-10"
      >
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            -
          </span>
        </span>
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            1
          </span>
        </span>
        <span
          class="dcloud-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="dcloud-scroll-number-only-unit current"
          >
            0
          </span>
        </span>
      </sup>
    </span>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge rtl render component should be rendered correctly in RTL direction 1`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper dcloud-badge-rtl"
  />,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge rtl render component should be rendered correctly in RTL direction 2`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-rtl"
  >
    <a
      class="head-example"
      href="#"
    >
      head
    </a>
    <sup
      class="dcloud-scroll-number dcloud-badge-count"
      data-show="true"
      style="margin-top: 10px; left: 10px;"
      title="5"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          5
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should be compatible with borderColor style 1`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count"
      data-show="true"
      style="background-color: rgb(255, 255, 255); color: rgb(153, 153, 153); border-color: #d9d9d9; box-shadow: 0 0 0 1px #d9d9d9 inset;"
      title="4"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          4
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should render when count is changed 1`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
      data-show="true"
      title="10"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          1
        </span>
      </span>
      <span
        class="dcloud-scroll-number-only"
        style="transform: translateY(-100%);"
      >
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -900%; left: 0px;"
        >
          0
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -800%; left: 0px;"
        >
          1
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -700%; left: 0px;"
        >
          2
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -600%; left: 0px;"
        >
          3
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -500%; left: 0px;"
        >
          4
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -400%; left: 0px;"
        >
          5
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -300%; left: 0px;"
        >
          6
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -200%; left: 0px;"
        >
          7
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -100%; left: 0px;"
        >
          8
        </span>
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          9
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 100%; left: 0px;"
        >
          0
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should render when count is changed 2`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
      data-show="true"
      title="11"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          1
        </span>
      </span>
      <span
        class="dcloud-scroll-number-only"
        style="transform: translateY(-100%);"
      >
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -900%; left: 0px;"
        >
          1
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -800%; left: 0px;"
        >
          2
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -700%; left: 0px;"
        >
          3
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -600%; left: 0px;"
        >
          4
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -500%; left: 0px;"
        >
          5
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -400%; left: 0px;"
        >
          6
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -300%; left: 0px;"
        >
          7
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -200%; left: 0px;"
        >
          8
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -100%; left: 0px;"
        >
          9
        </span>
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          0
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 100%; left: 0px;"
        >
          1
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should render when count is changed 3`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
      data-show="true"
      title="11"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          1
        </span>
      </span>
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
          style=""
        >
          1
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should render when count is changed 4`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
      data-show="true"
      title="111"
    >
      99+
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should render when count is changed 5`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count dcloud-badge-multiple-words"
      data-show="true"
      title="10"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          1
        </span>
      </span>
      <span
        class="dcloud-scroll-number-only"
        style="transition: none;"
      >
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          0
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should render when count is changed 6`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge dcloud-badge-not-a-wrapper"
  >
    <sup
      class="dcloud-scroll-number dcloud-badge-count"
      data-show="true"
      title="9"
    >
      <span
        class="dcloud-scroll-number-only"
        style="transform: translateY(100%);"
      >
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: -100%; left: 0px;"
        >
          9
        </span>
        <span
          class="dcloud-scroll-number-only-unit current"
        >
          0
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 100%; left: 0px;"
        >
          1
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 200%; left: 0px;"
        >
          2
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 300%; left: 0px;"
        >
          3
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 400%; left: 0px;"
        >
          4
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 500%; left: 0px;"
        >
          5
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 600%; left: 0px;"
        >
          6
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 700%; left: 0px;"
        >
          7
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 800%; left: 0px;"
        >
          8
        </span>
        <span
          class="dcloud-scroll-number-only-unit"
          style="position: absolute; top: 900%; left: 0px;"
        >
          9
        </span>
      </span>
    </sup>
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Badge should support offset when count is a ReactNode 1`] = `
LoadedCheerio {
  "0": <span
    class="dcloud-badge"
  >
    <a
      class="head-example"
      href="#"
    >
      head
    </a>
    <span
      class="dcloud-scroll-number-custom-component custom"
      style="margin-top: 20px; right: -10px; color: rgb(245, 34, 45);"
    />
  </span>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
